### PR TITLE
[SQL-Gen] Implement column constraints

### PR DESCRIPTION
**Pull request by Lewiky**
Implement ability to put column level constraints on values in the database

Tested + Passing

[](https://media.giphy.com/media/LllA2dKt1qZuE/giphy.gif)

### CODE REVIEW

File: src/main/scala/generate/database/PostgresGenerator.scala [hunk 1]
```diff
@@ -6,17 +6,40 @@ import utils.StringUtils._
...
     indent(
       column match {
-        case IntColumn(name)        => s"$name INT"
-        case FloatColumn(name)      => s"$name REAL"
-        case StringColumn(name)     => s"$name TEXT"
-        case BoolColumn(name)       => s"$name BOOLEAN"
-        case DateColumn(name)       => s"$name DATE"
-        case TimeColumn(name)       => s"$name TIME"
-        case DateTimeTzColumn(name) => s"$name TIMESTAMPTZ"
+        case IntColumn(name, constraints)    => s"$name INT" + constraints.map(parseConstraint).map(" " + _).mkString
```
Comment (jaylees14) [inline src/main/scala/generate/database/PostgresGenerator.scala]: Maybe bring constraints up to the `trait` level, then perform the addition with constraints at the end?

**Review by jaylees14 — COMMENTED** [review]
Nice job 👍

File: src/main/scala/generate/database/PostgresGenerator.scala [hunk 2]
```diff
@@ -6,17 +6,40 @@ import utils.StringUtils._
...
+  /** Given a comparison, parse it into the Postgres format */
+  private def parseComparison(comparison: Comparison): String =
+    comparison match {
+      case GreaterEqual => ">="
+      case Greater      => ">"
+      case Equal        => "=="
+      case NotEqual     => "!="
+      case Less         => "<"
+      case LessEqual    => "<="
+    }
```
Comment (c-harding) [inline src/main/scala/generate/database/PostgresGenerator.scala]: This is not parsing

File: src/main/scala/generate/database/PostgresGenerator.scala [hunk 3]
```diff
@@ -6,17 +6,40 @@ import utils.StringUtils._
...
+  /** Given a column constraint, parse it into the Postgres format */
+  private def parseConstraint(constraint: ColumnConstraint): String =
+    constraint match {
+      case NonNull                    => "NOT NULL"
+      case Null                       => "NULL"
+      case Check(left, comp, right)   => s"CHECK ($left " + parseComparison(comp) + s" $right)"
+      case Unique                     => "UNIQUE"
+      case PrimaryKey                 => "PRIMARY KEY"
+      case References(table, colName) => s"REFERENCES $table($colName)"
+    }
```
Comment (c-harding) [inline src/main/scala/generate/database/PostgresGenerator.scala]: This is not parsing

File: src/main/scala/generate/database/PostgresGenerator.scala [hunk 4]
```diff
@@ -6,17 +6,40 @@ import utils.StringUtils._
...
+
   /** Given a column, parse it into the type required by Postgres */
   private def parseColumn(column: Column): String =
```
Comment (c-harding) [inline src/main/scala/generate/database/PostgresGenerator.scala]: Ditto function name

File: src/main/scala/generate/database/ast/Column.scala [hunk 5]
```diff
@@ -1,20 +1,18 @@
...
+case class BoolColumn(name: String, constraints: List[ColumnConstraint] = List()) extends Column
 
-case class DateColumn(name: String) extends Column
+case class DateColumn(name: String, constraints: List[ColumnConstraint] = List()) extends Column
 
-case class TimeColumn(name: String) extends Column
+case class TimeColumn(name: String, constraints: List[ColumnConstraint] = List()) extends Column
 
-case class DateTimeTzColumn(name: String) extends Column
+case class DateTimeTzColumn(name: String, constraints: List[ColumnConstraint] = List()) extends Column
```
Comment (c-harding) [inline src/main/scala/generate/database/ast/Column.scala]: Adopt to Will’s refactoring

File: src/main/scala/generate/database/ast/ColumnConstraint.scala [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+case object Null extends ColumnConstraint
+
+case class Check(left: String, comp: Comparison, right: String) extends ColumnConstraint
+
+case object Unique extends ColumnConstraint
+
+case object PrimaryKey extends ColumnConstraint
+
+case class References(table: String, colName: String) extends ColumnConstraint
```
Comment (c-harding) [inline src/main/scala/generate/database/ast/ColumnConstraint.scala]: Personally not a fan of all the newlines…

File: src/main/scala/generate/database/ast/Comparison.scala [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+case object Greater extends Comparison
+
+case object Equal extends Comparison
+
+case object NotEqual extends Comparison
+
+case object Less extends Comparison
+
+case object LessEqual extends Comparison
```
Comment (c-harding) [inline src/main/scala/generate/database/ast/Comparison.scala]: Personally not a fan of all the newlines…

**Review by c-harding — COMMENTED** [review]
Only minor things

File: src/main/scala/generate/database/PostgresGenerator.scala (hunk 2 above)
Comment (Lewiky) [reply src/main/scala/generate/database/PostgresGenerator.scala]: You are not parsing

File: src/main/scala/generate/database/PostgresGenerator.scala (hunk 3 above)
Comment (Lewiky) [reply src/main/scala/generate/database/PostgresGenerator.scala]: You are not parsing

File: src/main/scala/generate/database/ast/Statement.scala [hunk 8]
```diff
@@ -4,4 +4,5 @@ package generate.database.ast
 sealed trait Statement
 
 case class Create(tableName: String, columns: List[ColumnDef]) extends Statement
-case class Read(tableName: String, columns: List[Column])      extends Statement
+
+case class Read(tableName: String, columns: List[Column]) extends Statement
```
Comment (jaylees14) [inline src/main/scala/generate/database/ast/Statement.scala]: Wrap this in a `Statement` object and drop the newlines

File: src/main/scala/generate/database/PostgresGenerator.scala [hunk 9]
```diff
@@ -1,12 +1,37 @@
...
+      case NonNull                    => "NOT NULL"
+      case Null                       => "NULL"
+      case Check(left, comp, right)   => s"CHECK ($left " + parseComparison(comp) + s" $right)"
+      case Unique                     => "UNIQUE"
+      case PrimaryKey                 => "PRIMARY KEY"
+      case References(table, colName) => s"REFERENCES $table($colName)"
+    }
+
+  @inline private def parseColumnConstraints(constraints: List[ColumnConstraint]): String =
+    constraints.map(parseConstraint).mkString(" ", " ", "", "")
```
Comment (c-harding) [inline src/main/scala/generate/database/PostgresGenerator.scala]:
```scala
constraints.map(parseConstraint).map(" "+_).mkString
```

**Review by jaylees14 — COMMENTED** [review]
Just a small comment - otherwise agree with @xsanda's existing comments. Looks good though! 👍

File: src/main/scala/generate/database/ast/Comparison.scala [hunk 10]
```diff
@@ -0,0 +1,20 @@
+package generate.database.ast
+
+/** AST representation for boolean comparisons */
+sealed trait Comparison
+
+object Comparison {
+
```
Comment (c-harding) [inline src/main/scala/generate/database/ast/Comparison.scala]: can we remove all these spaces?

File: src/main/scala/generate/database/ast/Statement.scala (hunk 8 above)
Comment (c-harding) [reply src/main/scala/generate/database/ast/Statement.scala]: as @jaylees14 said

**Review by c-harding — APPROVED** [review]
<3

**Review by jaylees14 — APPROVED** [review]

**Pull request merged**